Fix crate name validation

fmease opened the pull request:
Reject macro calls inside attribute `#![crate_name]` like in `#![crate_name = concat!("na", "me")]`.

Prior to #117584, the result of the expansion (here: `"name"`) would actually be properly picked up by the compiler and used as the crate name. However since #117584 / on master, we extract the "value" (i.e., the *literal* string literal) of the `#![crate_name]` much earlier in the pipeline way before macro expansion and **skip**/**ignore** any `#![crate_name]`s "assigned to" a macro call. See also #122001.

T-lang has ruled to reject `#![crate_name = MACRO!(...)]` outright very similar to other built-in attributes whose value we need early like `#![crate_type]`. See accepted FCP: https://github.com/rust-lang/rust/issues/122001#issuecomment-2023203182.

Note that the check as implemented in this PR is even more "aggressive" compared to the one of `#![crate_type]` by running as early as possible in order to reject `#![crate_name = MACRO!(...)]` even in "non-normal" executions of `rustc`, namely on *print requests* (e.g., `--print=crate-name` and `--print=file-names`). If I were to move the validation step a bit further back close to the `#![crate_type]` one, `--print=crate-name` (etc.) would *not* exit fatally with an error in this kind of situation but happily report an incorrect crate name (i.e., the "crate name" as if `#![crate_name]` didn't exist / deduced from other sources like `--crate-name` or the file name) which would match the behavior on master. Again, see also #122001.

I'm mentioning this explicitly because I'm not sure if it was that clear in the FCP'ed issue. I argue that my current approach is the most reasonable one. I know (from reading the code and from past experiments) that various print requests are still quite broken (mostly lack of validation).

To the best of my knowledge, there's no print request whose output references/contains a crate *type*, so there's no "inherent need" to move `#![crate_type]`'s validation to happen earlier.

---

Fixes #122001.

https://github.com/rust-lang/rust/labels/relnotes: Compatibility. Breaking change.

fmease committed:
Clean up rustc_session::output::{find,validate}_crate_name

fmease committed:
Move `#![crate_type]` UI tests into attributes directory

Gets rid of two top-level UI tests which is always great.

Furthermore, move `need-crate-arg-ignore-tidy$x.rs`
from `command/` to `invalid-compile-flags/`.
`command/` concerns `std::process::Command` tests, not CLI tests.

fmease committed:
Reject macro calls inside of `#![crate_name]`

fmease commented on compiler/rustc_interface/src/util.rs:
```diff
@@ -504,15 +513,14 @@ pub fn build_output_filenames(attrs: &[ast::Attribute], sess: &Session) -> Outpu
                 sess.dcx().emit_warn(errors::IgnoringOutDir);
             }
 
-            let out_filestem =
-                out_file.filestem().unwrap_or_default().to_str().unwrap().to_string();
+            let out_filestem = out_file.filestem().unwrap_or_default().to_str().unwrap();
             OutputFilenames::new(
                 out_file.parent().unwrap_or_else(|| Path::new("")).to_path_buf(),
-                crate_name.unwrap_or_else(|| out_filestem.replace('-', "_")),
+                crate_name,
```
Indeed, this is a change in behavior. On master, the "crate name" falls back to the file stem of the explicit outfile (`-o`) which is not correct (it doesn't agree with `tcx.crate_name(LOCAL_CRATE)`) (this happens since #114750). Note that this "crate name" is only used when emitting metadata files via `--emit=metadata` namely inside `libCRATE_NAME.rmeta`.

However, I couldn't actually come up with an actual reproducer as `--emit=metadata` doesn't actually emit anything if an `-o OUTPATH` is present **unless** an outpath was explicitly specified via `--emit=metadata=RMETA_OUTPATH`. So it might not be reachable at all and thus have no user-facing impact.

I don't know if `--emit=metadata -o OUTPATH` not emitting anything is intentional. Needs investigation (by me).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9878239644/job/27281740252) [(plain)](https://github.com/rust-lang/rust/commit/7b933ade27bcf87e571f0b7624c6a07601d78491/checks/27281740252/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
diff of stderr:

2   --> $SRC_DIR/core/src/iter/adapters/filter.rs:LL:COL
3    |
4    = help: consider adding a `#![type_length_limit="20156994"]` attribute to your crate
-    = note: the full type name has been written to '$TEST_BUILD_DIR/codegen/overflow-during-mono/overflow-during-mono.long-type.txt'
+    = note: the full type name has been written to '$TEST_BUILD_DIR/codegen/overflow-during-mono/overflow_during_mono.long-type.txt'
7 error: aborting due to 1 previous error
8 



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/codegen/overflow-during-mono/overflow-during-mono.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args codegen/overflow-during-mono.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/codegen/overflow-during-mono.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/codegen/overflow-during-mono" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/codegen/overflow-during-mono/auxiliary"
--- stderr -------------------------------
--- stderr -------------------------------
error: reached the type-length limit while instantiating `<Filter<Filter<std::array::IntoIter<i32, 11>, {closure@/checkout/tests/ui/codegen/overflow-during-mono.rs:13:41: 13:44}>, ...> as Iterator>::try_fold::<..., ..., ...>`
   |
   |
   = help: consider adding a `#![type_length_limit="20156994"]` attribute to your crate
   = note: the full type name has been written to '/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/codegen/overflow-during-mono/overflow_during_mono.long-type.txt'
error: aborting due to 1 previous error
------------------------------------------


---
---- [ui] tests/ui/diagnostic-width/long-E0308.rs stdout ----
diff of stderr:

20    |
21    = note: expected struct `Atype<Btype<..., ...>, ...>`
22                 found enum `Result<Result<..., ...>, ...>`
-    = note: the full type name has been written to '$TEST_BUILD_DIR/diagnostic-width/long-E0308/long-E0308.long-type-hash.txt'
+    = note: the full type name has been written to '$TEST_BUILD_DIR/diagnostic-width/long-E0308/long_E0308.long-type-hash.txt'
24    = note: consider using `--verbose` to print the full type name to the console
26 error[E0308]: mismatched types

36    |
36    |
37    = note: expected enum `Option<Result<..., ...>>`
38               found enum `Result<Result<..., ...>, ...>`
-    = note: the full type name has been written to '$TEST_BUILD_DIR/diagnostic-width/long-E0308/long-E0308.long-type-hash.txt'
+    = note: the full type name has been written to '$TEST_BUILD_DIR/diagnostic-width/long-E0308/long_E0308.long-type-hash.txt'
40    = note: consider using `--verbose` to print the full type name to the console
42 error[E0308]: mismatched types

56    |
56    |
57    = note: expected struct `Atype<Btype<..., ...>, ...>`
58            found unit type `()`
-    = note: the full type name has been written to '$TEST_BUILD_DIR/diagnostic-width/long-E0308/long-E0308.long-type-hash.txt'
+    = note: the full type name has been written to '$TEST_BUILD_DIR/diagnostic-width/long-E0308/long_E0308.long-type-hash.txt'
60    = note: consider using `--verbose` to print the full type name to the console
62 error[E0308]: mismatched types

74    |
75    = note: expected unit type `()`
75    = note: expected unit type `()`
76                    found enum `Result<Result<..., ...>, ...>`
-    = note: the full type name has been written to '$TEST_BUILD_DIR/diagnostic-width/long-E0308/long-E0308.long-type-hash.txt'
+    = note: the full type name has been written to '$TEST_BUILD_DIR/diagnostic-width/long-E0308/long_E0308.long-type-hash.txt'
78    = note: consider using `--verbose` to print the full type name to the console
80 error: aborting due to 4 previous errors


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/diagnostic-width/long-E0308/long-E0308.stderr
To only update this specific test, also pass `--test-args diagnostic-width/long-E0308.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/diagnostic-width/long-E0308.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/diagnostic-width/long-E0308" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/diagnostic-width/long-E0308/auxiliary" "--diagnostic-width=60" "-Zwrite-long-types-to-disk=yes"
--- stderr -------------------------------
error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/diagnostic-width/long-E0308.rs:44:9
   |
   |
LL |        let x: Atype<
   |  _____________-
LL | |        Btype<
LL | |          Ctype<
LL | |            Atype<
...  |
LL | |        i32
LL | |      > = Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok...
   | | _____-___^
   |  |     expected due to this
   |  |     expected due to this
LL |  |         Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok...
LL |  |             Ok("")
LL |  |         ))))))))))))))))))))))))))))))
LL |  |     ))))))))))))))))))))))))))))));
   |  |__________________________________^ expected `Atype<Btype<..., ...>, ...>`, found `Result<Result<..., ...>, ...>`
   |
   = note: expected struct `Atype<Btype<..., ...>, ...>`
                found enum `Result<Result<..., ...>, ...>`
   = note: the full type name has been written to '/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/diagnostic-width/long-E0308/long_E0308.long-type-922878406246076682.txt'
   = note: consider using `--verbose` to print the full type name to the console
error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/diagnostic-width/long-E0308.rs:57:26
   |
   |
LL |       ))))))))))))))))) == Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(O...
   |  __________________________^
LL | |         Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(...
LL | |             Ok(Ok(Ok(Ok(Ok(Ok(Ok("")))))))
LL | |         ))))))))))))))))))))))))))))))
LL | |     ))))))))))))))))))))))));
   | |____________________________^ expected `Option<Result<..., ...>>`, found `Result<Result<..., ...>, ...>`
   |
   = note: expected enum `Option<Result<..., ...>>`
              found enum `Result<Result<..., ...>, ...>`
   = note: the full type name has been written to '/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/diagnostic-width/long-E0308/long_E0308.long-type-5722041322995796003.txt'
   = note: consider using `--verbose` to print the full type name to the console
error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/diagnostic-width/long-E0308.rs:88:9
   |
LL |       let x: Atype<
LL |       let x: Atype<
   |  ____________-
LL | |       Btype<
LL | |         Ctype<
LL | |           Atype<
...  |
LL | |       i32
LL | |     > = ();
   | |     -   ^^ expected `Atype<Btype<..., ...>, ...>`, found `()`
   |       expected due to this
   |
   |
   = note: expected struct `Atype<Btype<..., ...>, ...>`
           found unit type `()`
   = note: the full type name has been written to '/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/diagnostic-width/long-E0308/long_E0308.long-type-922878406246076682.txt'
   = note: consider using `--verbose` to print the full type name to the console
error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/diagnostic-width/long-E0308.rs:91:17
   |
   |
LL |       let _: () = Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(O...
   |  ____________--___^
   | |            expected due to this
   | |            expected due to this
LL | |         Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(Ok(...
LL | |             Ok(Ok(Ok(Ok(Ok(Ok(Ok("")))))))
LL | |         ))))))))))))))))))))))))))))))
LL | |     ))))))))))))))))))))))));
   | |____________________________^ expected `()`, found `Result<Result<..., ...>, ...>`
   = note: expected unit type `()`
   = note: expected unit type `()`
                   found enum `Result<Result<..., ...>, ...>`
   = note: the full type name has been written to '/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/diagnostic-width/long-E0308/long_E0308.long-type-11912194927516380970.txt'
   = note: consider using `--verbose` to print the full type name to the console
error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0308`.
------------------------------------------
------------------------------------------


---- [ui] tests/ui/infinite/infinite-instantiation.rs stdout ----
diff of stderr:

9    |
10 LL | fn function<T:ToOpt + Clone>(counter: usize, t: T) {
11    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-    = note: the full type name has been written to '$TEST_BUILD_DIR/infinite/infinite-instantiation/infinite-instantiation.long-type.txt'
+    = note: the full type name has been written to '$TEST_BUILD_DIR/infinite/infinite-instantiation/infinite_instantiation.long-type.txt'
14 error: aborting due to 1 previous error
15 



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/infinite/infinite-instantiation/infinite-instantiation.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args infinite/infinite-instantiation.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/infinite/infinite-instantiation.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/infinite/infinite-instantiation" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/infinite/infinite-instantiation/auxiliary"
--- stderr -------------------------------
--- stderr -------------------------------
error: reached the recursion limit while instantiating `function::<Option<Option<Option<Option<Option<...>>>>>>`
   |
   |
LL |         function(counter - 1, t.to_option());
   |
note: `function` defined here
  --> /checkout/tests/ui/infinite/infinite-instantiation.rs:20:1
   |
   |
LL | fn function<T:ToOpt + Clone>(counter: usize, t: T) {
   = note: the full type name has been written to '/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/infinite/infinite-instantiation/infinite_instantiation.long-type.txt'

error: aborting due to 1 previous error
------------------------------------------
------------------------------------------


---- [ui] tests/ui/issues/issue-37311-type-length-limit/issue-37311.rs stdout ----
diff of stderr:

5    |         ^^^^^^^^^^^^^^^^^^^^^^
6    |
7    = help: consider adding a `#![type_length_limit="33554429"]` attribute to your crate
-    = note: the full type name has been written to '$TEST_BUILD_DIR/issues/issue-37311-type-length-limit/issue-37311/issue-37311.long-type.txt'
+    = note: the full type name has been written to '$TEST_BUILD_DIR/issues/issue-37311-type-length-limit/issue-37311/issue_37311.long-type.txt'
10 error: aborting due to 1 previous error
11 



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-37311-type-length-limit/issue-37311/issue-37311.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args issues/issue-37311-type-length-limit/issue-37311.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/issues/issue-37311-type-length-limit/issue-37311.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-37311-type-length-limit/issue-37311" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-37311-type-length-limit/issue-37311/auxiliary"
--- stderr -------------------------------
--- stderr -------------------------------
error: reached the type-length limit while instantiating `<(&(&(..., ...), ...), ...) as Foo>::recurse`
   |
   |
LL |         (self, self).recurse();
   |
   |
   = help: consider adding a `#![type_length_limit="33554429"]` attribute to your crate
   = note: the full type name has been written to '/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-37311-type-length-limit/issue-37311/issue_37311.long-type.txt'
error: aborting due to 1 previous error
------------------------------------------



---- [ui] tests/ui/issues/issue-8727.rs stdout ----
diff of stderr:

20    |
21 LL | fn generic<T>() {
22    | ^^^^^^^^^^^^^^^
-    = note: the full type name has been written to '$TEST_BUILD_DIR/issues/issue-8727/issue-8727.long-type.txt'
+    = note: the full type name has been written to '$TEST_BUILD_DIR/issues/issue-8727/issue_8727.long-type.txt'
25 error: aborting due to 1 previous error; 1 warning emitted
26 



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-8727/issue-8727.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args issues/issue-8727.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/issues/issue-8727.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-8727" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-8727/auxiliary"
--- stderr -------------------------------
warning: function cannot return without recursing
##[warning]  --> /checkout/tests/ui/issues/issue-8727.rs:7:1
   |
   |
LL | fn generic<T>() { //~ WARN function cannot return without recursing
   | ^^^^^^^^^^^^^^^ cannot return without recursing
LL |     generic::<Option<T>>();
   |     ---------------------- recursive call site
   |
   = help: a `loop` may express intention better if this is on purpose
   = note: `#[warn(unconditional_recursion)]` on by default

error: reached the recursion limit while instantiating `generic::<Option<Option<Option<Option<Option<...>>>>>>`
   |
LL |     generic::<Option<T>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^
   |
   |
note: `generic` defined here
  --> /checkout/tests/ui/issues/issue-8727.rs:7:1
   |
LL | fn generic<T>() { //~ WARN function cannot return without recursing
   | ^^^^^^^^^^^^^^^
   = note: the full type name has been written to '/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-8727/issue_8727.long-type.txt'
error: aborting due to 1 previous error; 1 warning emitted
------------------------------------------



---- [ui] tests/ui/issues/issue-67552.rs stdout ----
diff of stderr:

11 LL | | where
12 LL | |     T: Iterator,
13    | |________________^
-    = note: the full type name has been written to '$TEST_BUILD_DIR/issues/issue-67552/issue-67552.long-type.txt'
+    = note: the full type name has been written to '$TEST_BUILD_DIR/issues/issue-67552/issue_67552.long-type.txt'
16 error: aborting due to 1 previous error
17 



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-67552/issue-67552.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args issues/issue-67552.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/issues/issue-67552.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-67552" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-67552/auxiliary" "-Copt-level=0"
--- stderr -------------------------------
--- stderr -------------------------------
error: reached the recursion limit while instantiating `rec::<&mut &mut &mut &mut &mut ...>`
   |
   |
LL |         rec(identity(&mut it))
   |
note: `rec` defined here
  --> /checkout/tests/ui/issues/issue-67552.rs:22:1
   |
   |
LL | / fn rec<T>(mut it: T)
LL | | where
LL | |     T: Iterator,
   | |________________^
   = note: the full type name has been written to '/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-67552/issue_67552.long-type.txt'
error: aborting due to 1 previous error
------------------------------------------



---- [ui] tests/ui/iterators/issue-58952-filter-type-length.rs stdout ----
diff of stderr:

2   --> $SRC_DIR/core/src/iter/adapters/filter.rs:LL:COL
3    |
4    = help: consider adding a `#![type_length_limit="21233607"]` attribute to your crate
-    = note: the full type name has been written to '$TEST_BUILD_DIR/iterators/issue-58952-filter-type-length/issue-58952-filter-type-length.long-type.txt'
+    = note: the full type name has been written to '$TEST_BUILD_DIR/iterators/issue-58952-filter-type-length/issue_58952_filter_type_length.long-type.txt'
7 error: aborting due to 1 previous error
8 



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/iterators/issue-58952-filter-type-length/issue-58952-filter-type-length.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args iterators/issue-58952-filter-type-length.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/iterators/issue-58952-filter-type-length.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/iterators/issue-58952-filter-type-length" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/iterators/issue-58952-filter-type-length/auxiliary"
--- stderr -------------------------------
--- stderr -------------------------------
error: reached the type-length limit while instantiating `<std::vec::IntoIter<i32> as Iterator>::try_fold::<vec::in_place_drop::InPlaceDrop<i32>, {closure@iter::adapters::filter::filter_try_fold<'_, ..., ..., ..., ..., ...>::{closure#0}}, ...>`
   |
   |
   = help: consider adding a `#![type_length_limit="21233607"]` attribute to your crate
   = note: the full type name has been written to '/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/iterators/issue-58952-filter-type-length/issue_58952_filter_type_length.long-type.txt'
error: aborting due to 1 previous error
------------------------------------------



---- [ui] tests/ui/lto/lto-duplicate-symbols.rs stdout ----
diff of stderr:

1 warning: Linking globals named 'foo': symbol multiply defined!
2 
- error: failed to load bitcode of module "lto-duplicate-symbols2.lto_duplicate_symbols2.HASH-cgu.0.rcgu.o": 
+ error: failed to load bitcode of module "lto_duplicate_symbols2.lto_duplicate_symbols2.b528957825d94e1e-cgu.0.rcgu.o": 
5 error: aborting due to 1 previous error; 1 warning emitted
6 



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lto/lto-duplicate-symbols/lto-duplicate-symbols.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args lto/lto-duplicate-symbols.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lto/lto-duplicate-symbols.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lto/lto-duplicate-symbols" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lto/lto-duplicate-symbols/auxiliary" "-C" "lto"
--- stderr -------------------------------
--- stderr -------------------------------
warning: Linking globals named 'foo': symbol multiply defined!

error: failed to load bitcode of module "lto_duplicate_symbols2.lto_duplicate_symbols2.b528957825d94e1e-cgu.0.rcgu.o": 
error: aborting due to 1 previous error; 1 warning emitted
------------------------------------------



---- [ui] tests/ui/recursion/issue-83150.rs stdout ----
diff of stderr:

17    |               ^^^^^^^^^^^^^^^^^^^
18    |
19    = help: consider adding a `#![type_length_limit="23068663"]` attribute to your crate
-    = note: the full type name has been written to '$TEST_BUILD_DIR/recursion/issue-83150/issue-83150.long-type.txt'
+    = note: the full type name has been written to '$TEST_BUILD_DIR/recursion/issue-83150/issue_83150.long-type.txt'
22 error: aborting due to 1 previous error; 1 warning emitted
23 



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/recursion/issue-83150/issue-83150.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args recursion/issue-83150.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/recursion/issue-83150.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/recursion/issue-83150" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/recursion/issue-83150/auxiliary" "-Copt-level=0"
--- stderr -------------------------------
warning: function cannot return without recursing
##[warning]  --> /checkout/tests/ui/recursion/issue-83150.rs:11:1
   |
   |
LL | fn func<T: Iterator<Item = u8>>(iter: &mut T) {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot return without recursing
LL |     //~^ WARN function cannot return without recursing
LL |     func(&mut iter.map(|x| x + 1))
   |     ------------------------------ recursive call site
   |
   = help: a `loop` may express intention better if this is on purpose
   = note: `#[warn(unconditional_recursion)]` on by default

error: reached the type-length limit while instantiating `<&mut Map<&mut Map<&mut ..., ...>, ...> as Iterator>::map::<..., ...>`
   |
LL |     func(&mut iter.map(|x| x + 1))
   |               ^^^^^^^^^^^^^^^^^^^
   |
   |
   = help: consider adding a `#![type_length_limit="23068663"]` attribute to your crate
   = note: the full type name has been written to '/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/recursion/issue-83150/issue_83150.long-type.txt'
error: aborting due to 1 previous error; 1 warning emitted
------------------------------------------



---- [ui] tests/ui/runtime/on-broken-pipe/child-processes.rs#error stdout ----

error in revision `error`: test run failed!
status: exit status: 101
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/runtime/on-broken-pipe/child-processes.error" && RUST_TEST_THREADS="16" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/runtime/on-broken-pipe/child-processes.error/a"
--- stderr -------------------------------
thread 'main' panicked at /checkout/tests/ui/runtime/on-broken-pipe/child-processes.rs:46:14:
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
------------------------------------------


---- [ui] tests/ui/runtime/on-broken-pipe/child-processes.rs#kill stdout ----

error in revision `kill`: test run failed!
status: exit status: 101
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/runtime/on-broken-pipe/child-processes.kill" && RUST_TEST_THREADS="16" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/runtime/on-broken-pipe/child-processes.kill/a"
--- stderr -------------------------------
thread 'main' panicked at /checkout/tests/ui/runtime/on-broken-pipe/child-processes.rs:46:14:
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
------------------------------------------


---- [ui] tests/ui/runtime/on-broken-pipe/child-processes.rs#inherit stdout ----

error in revision `inherit`: test run failed!
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/runtime/on-broken-pipe/child-processes.inherit" && RUST_TEST_THREADS="16" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/runtime/on-broken-pipe/child-processes.inherit/a"
stdout: none
--- stderr -------------------------------
thread 'main' panicked at /checkout/tests/ui/runtime/on-broken-pipe/child-processes.rs:46:14:
thread 'main' panicked at /checkout/tests/ui/runtime/on-broken-pipe/child-processes.rs:46:14:
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
------------------------------------------


---- [ui] tests/ui/runtime/on-broken-pipe/child-processes.rs#default stdout ----

error in revision `default`: test run failed!
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/runtime/on-broken-pipe/child-processes.default" && RUST_TEST_THREADS="16" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/runtime/on-broken-pipe/child-processes.default/a"
stdout: none
--- stderr -------------------------------
thread 'main' panicked at /checkout/tests/ui/runtime/on-broken-pipe/child-processes.rs:46:14:
---
diff of stderr:

20    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
21    |
22    = help: consider adding a `#![type_length_limit="27262965"]` attribute to your crate
-    = note: the full type name has been written to '$TEST_BUILD_DIR/traits/issue-91949-hangs-on-recursion/issue-91949-hangs-on-recursion.long-type.txt'
+    = note: the full type name has been written to '$TEST_BUILD_DIR/traits/issue-91949-hangs-on-recursion/issue_91949_hangs_on_recursion.long-type.txt'
25 error: aborting due to 1 previous error; 1 warning emitted
26 



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/issue-91949-hangs-on-recursion/issue-91949-hangs-on-recursion.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args traits/issue-91949-hangs-on-recursion.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/traits/issue-91949-hangs-on-recursion.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/issue-91949-hangs-on-recursion" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/issue-91949-hangs-on-recursion/auxiliary" "-Zinline-mir=no"
--- stderr -------------------------------
warning: function cannot return without recursing
##[warning]  --> /checkout/tests/ui/traits/issue-91949-hangs-on-recursion.rs:20:1
   |
   |
LL | / fn recurse<T>(elements: T) -> Vec<char>
LL | | //~^ WARN function cannot return without recursing
LL | |     T: Iterator<Item = ()>,
   | |___________________________^ cannot return without recursing
LL |   {
LL |   {
LL |       recurse(IteratorOfWrapped(elements).map(|t| t.0))
   |       ------------------------------------------------- recursive call site
   |
   = help: a `loop` may express intention better if this is on purpose
   = note: `#[warn(unconditional_recursion)]` on by default

error: reached the type-length limit while instantiating `<IteratorOfWrapped<(), Map<IteratorOfWrapped<(), ...>, ...>> as Iterator>::map::<..., ...>`
   |
   |
LL |     recurse(IteratorOfWrapped(elements).map(|t| t.0))
   |
   |
   = help: consider adding a `#![type_length_limit="27262965"]` attribute to your crate
   = note: the full type name has been written to '/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/issue-91949-hangs-on-recursion/issue_91949_hangs_on_recursion.long-type.txt'
error: aborting due to 1 previous error; 1 warning emitted
------------------------------------------


```

</details>

fmease reviewed (commented):
[deleted]

fmease commented on compiler/rustc_interface/src/passes.rs:
```diff
@@ -1044,23 +1044,91 @@ pub(crate) fn start_codegen<'tcx>(
+/// inside of the attributes, such as in `#![name = expand!()]`. Normal attribute validation happens
+/// during semantic analysis via [`TyCtxt::check_mod_attrs`] which happens *after* macro expansion
+/// when such macro calls (here: `expand`) have already been expanded and we can no longer check for
+/// their presence.
```
and `span.from_expansion()` is not "reliable": `include_str!(…)` doesn't intro an expansion context for example.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9879048532/job/27284386711) [(plain)](https://github.com/rust-lang/rust/commit/3de58e6b25425448c0860f258740a7f439ba558d/checks/27284386711/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.73s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
............................

failures:

---- errors::verify_interface_crate_name_invalid_1 stdout ----
thread 'errors::verify_interface_crate_name_invalid_1' panicked at compiler/rustc_interface/src/errors.rs:16:10:
CrateNameInvalid: variable `s` not found (interface_crate_name_invalid)


failures:
    errors::verify_interface_crate_name_invalid_1
```

</details>

fmease commented:
I've finally added a PR description. Sorry about the delay ^^'.

bors commented:
:umbrella: The latest upstream changes (presumably #125443) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

wesleywiser reviewed (commented):
Is the last commit still WIP?

fmease commented:
> Is the last commit still WIP?

I've now dropped the WIP commit as I still need to investigate the implications of its changes. I'll submit it as a follow-up PR if everything works out.

I've adjusted the wording in the comment to no longer use the word 'currently'.

Ready for re-review. Nothing WIP left.

bors commented:
:umbrella: The latest upstream changes (presumably #129647) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #130724) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #131185) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #125579) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

fmease commented:
I've dropped the `tests/ui/{invalid-compile-flags → cli}/` change. See discussion in #133277.

bors commented:
:umbrella: The latest upstream changes (presumably #134052) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #134243) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bjorn3 commented on compiler/rustc_interface/src/passes.rs:
```diff
@@ -1162,23 +1163,83 @@ pub(crate) fn start_codegen<'tcx>(
+pub fn get_crate_name(sess: &Session, krate_attrs: &[ast::Attribute]) -> Symbol {
+    // We unconditionally validate all `#![crate_name]`s even if a crate name was
+    // set on the command line via `--crate-name` which we prioritize over the
+    // crate attributes. We perform the validation here instead of later to ensure
```
`--crate-name` isn't prioritized. Rather it and `#![crate_name]` must match.

bjorn3 commented:
r? @bjorn3

fmease commented:
I take the approval as an "r=me with nit addressed".
@bors r=bjorn3

bors commented:
:pushpin: Commit 9b6fd35738965ef3f246018fddc743b5e5cd8d2c has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 9b6fd35738965ef3f246018fddc743b5e5cd8d2c -->
<!-- homu: {"type":"Approved","sha":"9b6fd35738965ef3f246018fddc743b5e5cd8d2c","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
